File: semester_3/operating_system/src/c/pthread/pthread-mutex.c
/*
 * pthread/pthread-mutex.c --
 *
 *      A simple demonstration of pthread mutexes. This example
 *      lacks error handling code to keep it short.
 */

#define _POSIX_C_SOURCE 200809L

#include <pthread.h>

typedef struct {
    unsigned int counter;       /* shared counter */
    pthread_mutex_t mutex;      /* mutex protecting the counter */
} counter_t;

static void* work(void *arg)
{
    counter_t *c = (counter_t *) arg;
    
    (void) pthread_mutex_lock(&c->mutex);
    c->counter++;
    (void) pthread_mutex_unlock(&c->mutex);
    return NULL;
}

int main(int argc, char *argv[])
{
    pthread_t tids[argc];
    counter_t cnter = { .counter = 0, .mutex = PTHREAD_MUTEX_INITIALIZER };

    (void) argv;

    for (int i = 1; i < argc; i++) {
        (void) pthread_create(&tids[i], NULL, work, &cnter);
    }
    for (int i = 1; i < argc; i++) {
        (void) pthread_join(tids[i], NULL);
    }
    return 0;
}

File: semester_3/operating_system/src/rust/pthread/src/bin/pthread-mutex.rs
/*
 * pthread/src/bin/pthread-mutex.rs --
 *
 *      A simple demonstration of mutexes in Rust. This example lacks
 *      error handling code to keep it short.
 */

use std::env;
use std::sync::{Arc, Mutex};
use std::thread;

fn work(arc: Arc<Mutex<u64>>) {
    let mut c = arc.lock().unwrap();
    *c += 1;
}

fn main() {
    let counter = Mutex::new(0);
    let arc = Arc::new(counter);

    let mut vec = Vec::new();
    for _ in env::args().skip(1) {
        let a = arc.clone();
        vec.push(thread::spawn(move || work(a)));
    }

    for handle in vec {
        handle.join().unwrap();
    }
}
